neuralmagic/benchmarks/scripts/benchmark_serving.py

```diff
@@ -0,0 +1,398 @@
+"""Benchmark online serving throughput.
+
+On the server side, run one of the following commands:
+    (vLLM backend)
+    python -m vllm.entrypoints.api_server \
+        --model <your_model> --swap-space 16 \
+        --disable-log-requests
+
+    (TGI backend)
+    ./launch_hf_server.sh <your_model>
+
+On the client side, run:
+    python benchmarks/benchmark_serving.py \
+        --backend <backend> \
+        --tokenizer <your_model> --dataset <target_dataset> \
+        --request-rate <request_rate>
+
+NOTE: This script is copied from upstream vllm repo (february 13th, 2024).
+"""
+import argparse
+import asyncio
+import json
+import random
+import time
+from pathlib import Path
+from dataclasses import dataclass
+from datetime import datetime
+from typing import AsyncGenerator, List, Tuple
+
+import numpy as np
+from tqdm.asyncio import tqdm
+from transformers import PreTrainedTokenizerBase
+from vllm.transformers_utils.tokenizer import get_tokenizer
+from common import get_bench_environment, sample_requests, generate_synthetic_requests
+
+from backend_request_func import (
+    ASYNC_REQUEST_FUNCS,
+    RequestFuncInput,
+    RequestFuncOutput,
+)
+
+
+@dataclass
+class BenchmarkMetrics:
+    completed: int
+    total_input: int
+    total_output: int
+    request_throughput: float
+    input_throughput: float
+    output_throughput: float
+    median_request_latency: float
+    p90_request_latency: float
+    p99_request_latency: float
+    mean_ttft_ms: float
+    median_ttft_ms: float
+    p90_ttft_ms: float
+    p99_ttft_ms: float
+    mean_tpot_ms: float
+    median_tpot_ms: float
+    p90_tpot_ms: float
+    p99_tpot_ms: float
+
+
+async def get_request(
+    input_requests: List[Tuple[str, int, int]],
+    request_rate: float,
+) -> AsyncGenerator[Tuple[str, int, int], None]:
+    input_requests = iter(input_requests)
+    for request in input_requests:
+        yield request
+
+        if request_rate == float("inf"):
+            # If the request rate is infinity, then we don't need to wait.
+            continue
+        # Sample the request interval from the exponential distribution.
+        interval = np.random.exponential(1.0 / request_rate)
+        # The next request will be sent after the interval.
+        await asyncio.sleep(interval)
+
+
+def calculate_metrics(
+    input_requests: List[Tuple[str, int, int]],
+    outputs: List[RequestFuncOutput],
+    dur_s: float,
+    tokenizer: PreTrainedTokenizerBase,
+) -> BenchmarkMetrics:
+    total_output = 0
+    total_input = 0
+    completed = 0
+    latencies = []
+    per_token_latencies = []
+    ttfts = []
+    for i in range(len(outputs)):
+        if outputs[i].success:
+            output_len = len(tokenizer.encode(outputs[i].generated_text))
+            total_output += output_len
+            total_input += input_requests[i][1]
+            latencies.append(outputs[i].latency)
+            per_token_latencies.append(outputs[i].latency / output_len)
```

Comment:
Note: this is not actually computing TPOT.

It is computing `avg_e2e_latency / number_of_tokens_generated`

A true TPOT analysis would look at the distribution of to time for each token.

We can compute `average_tpot` at the request level with:
- `per_token_latencies.append((outputs[i].latency - outputs[i].ttft) / output_len)`

When we compute the P50, P90, P99 of these, we will be computing the distribution of AVERAGE TPOT at the request level.

Another interesting metric to consider would be the distribution of TPOT at the token level. This will require us to change the vLLM generate server to stream tokens rather than text. Let's tackle this another time

Reply:
Fixed . Thanks for catching this @robertgshaw2-neuralmagic !